src/internal/group_api/mod.rs

```diff
@@ -307,37 +360,78 @@ pub fn group_create<'a, CR: rand::CryptoRng + rand::RngCore>(
-                } else {
-                    None
-                };
-                (group_pub_key, encrypted_group_key, transform_key)
-            })
+) -> impl Future<Item = GroupCreateResult, Error = IronOxideErr> + 'a {
+    user_api::user_key_list(auth, members)
+        .and_then(move |member_ids_and_keys| {
+            // this will occur when one of the UserIds in the members list cannot be found
+            if member_ids_and_keys.len() != members.len() {
+                // figure out which user ids could not be found in the database and include them in the error message.
+                use std::{collections::HashSet, iter::FromIterator};
+                let desired_members_set: HashSet<&UserId> = HashSet::from_iter(members);
+                let found_members: Vec<UserId> =
+                    member_ids_and_keys.into_iter().map(|(x, _)| x).collect();
+                let found_members_set: HashSet<&UserId> = HashSet::from_iter(&found_members);
+                // TODO (reviewers): this double reference scares me, but seems to work? Is there a better way to do this?
+                let diff: HashSet<&&UserId> =
+                    desired_members_set.difference(&found_members_set).collect();
+                futures::future::err(IronOxideErr::UserDoesNotExist(format!(
+                    "Failed to find the following users from the `members` list: {:?}",
+                    diff
+                )))
+            } else {
+                transform::gen_group_keys(recrypt)
+                    .and_then(move |(plaintext, group_priv_key, group_pub_key)| {
+                        // encrypt the group secret to the current user as they will be the group admin
+                        let encrypted_group_key = recrypt.encrypt(
+                            &plaintext,
+                            &user_master_pub_key.into(),
+                            &auth.signing_private_key().into(),
+                        )?;
+
+                        let maybe_member_info: Result<Vec<(UserId, PublicKey, TransformKey)>, _> =
+                            member_ids_and_keys
+                                .into_iter()
+                                .map(|(id, member_pub_key)| {
+                                    let maybe_transform_key = recrypt.generate_transform_key(
+                                        &group_priv_key.clone().into(),
+                                        &member_pub_key.clone().into(),
+                                        &auth.signing_private_key().into(),
+                                    );
+                                    maybe_transform_key.map(|transform_key| {
+                                        (id, member_pub_key, transform_key.into())
+                                    })
+                                })
+                                .collect();
+
+                        let mut member_info = maybe_member_info?;
+
+                        if add_as_member {
```

Comment:
It seems like you could just add the id and keys to the members list to avoid this duplication.

Reply:
I'm having issues modifying references to vectors in this function without the compiler eventually yelling at me, so this was my way to avoid looking up the public key with the others since we already know it. Hopefully we'll get the compiler sorted out.